commons/com.b2international.index/src/com/b2international/index/revision/RevisionConflictProcessor.java

```diff
@@ -91,12 +105,104 @@ default Conflict convertConflict(Conflict conflict) {
+
+				JsonDiff sourceDiff = JsonDiff.diff(oldObject, sourceObject).withoutNullAdditions(mapper);
+				JsonDiff targetDiff = JsonDiff.diff(oldObject, targetObject).withoutNullAdditions(mapper);
+				
+				Map<String, JsonChange> sourceChanges = Maps.uniqueIndex(sourceDiff.getChanges(), JsonChange::getRootFieldPath);
+				Map<String, JsonChange> targetChanges = Maps.uniqueIndex(targetDiff.getChanges(), JsonChange::getRootFieldPath);
+				
+				final Set<String> conflictingFields = Sets.intersection(sourceChanges.keySet(), targetChanges.keySet());
+				
+				for (String conflictingField : conflictingFields) {
+					JsonChange sourceJsonChange = sourceChanges.get(conflictingField);
+					JsonChange targetJsonChange = targetChanges.get(conflictingField);
+					
+					// report conflict if any change differs in any way on any property
+					if (!Objects.equals(sourceJsonChange, targetJsonChange)) {
+						return null;
+					}
+				}
+
+				// otherwise apply both changes on old object to merge them
+				sourceDiff.applyInPlace(oldObject);
+				targetDiff.applyInPlace(oldObject);
+				
+				// if the resulting object would be empty, then null it out instead
+				return sourceChange.withNewValue(isEmpty(oldObject) ? null : oldObject.toString());
+			} catch (JsonProcessingException e) {
+				throw new IndexException("Couldn't parse json", e);
+			}
+		}
+		
+		private boolean isEmpty(JsonNode node) {
+			if (node == null || node.isNull()) {
+				return true;
+			} else if (node.isContainerNode()) {
+				for (JsonNode child : node) {
+					if (!isEmpty(child)) {
+						return false;
+					}
+				}
+				return true;
+			} else {
+				return false;
+			}
+		}
+
+		protected RevisionPropertyDiff handleCollectionConflict(DocumentMapping mapping, RevisionPropertyDiff sourceChange, RevisionPropertyDiff targetChange, ObjectMapper mapper, boolean isSet) {
+			try {
+				ArrayNode oldArray = CompareUtils.isEmpty(sourceChange.getOldValue()) ? mapper.createArrayNode() : (ArrayNode) mapper.readTree(sourceChange.getOldValue());
+				ArrayNode sourceArray = CompareUtils.isEmpty(sourceChange.getNewValue()) ? mapper.createArrayNode() : (ArrayNode) mapper.readTree(sourceChange.getNewValue());
+				ArrayNode targetArray = CompareUtils.isEmpty(targetChange.getNewValue()) ? mapper.createArrayNode() : (ArrayNode) mapper.readTree(targetChange.getNewValue());
+				
+				ArrayDiff sourceDiff = ArrayDiff.diff(oldArray, sourceArray);
+				ArrayDiff targetDiff = ArrayDiff.diff(oldArray, targetArray);
+				
+				// TODO nested item conflict resolution???
+				// create a big union of all changes, treat collection as set if needed
+				// this eliminates most of the unnecessary conflicts between two collection properties
+				sourceDiff.getAddedItems().forEach(oldArray::add);
+				Set<JsonNode> newSourceItems = Set.copyOf(sourceDiff.getAddedItems());
+				targetDiff.getAddedItems().forEach(newTargetItem -> {
+					if (!isSet || !newSourceItems.contains(newTargetItem)) {
```

Comment:
Maybe I'm overly cautious, but to be _really_ sure that we uphold `Set` semantics, `oldArray` could be checked here instead of `newSourceItems`.

```suggestion
					if (!isSet || !oldArray.contains(newTargetItem)) {
```

Reply:
Not sure, `oldArray` is an ArrayNode and that does not have a contains method. Also, `sourceDiff` items need to be checked as well if we are overly cautious.